samples/nrf9160/modem_shell/src/cloud/cloud_shell.c

```diff
@@ -0,0 +1,192 @@
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <stdlib.h>
+#include <zephyr.h>
+#include <net/nrf_cloud.h>
+#include <shell/shell.h>
+#include "mosh_print.h"
+
+#if defined(CONFIG_NRF_CLOUD_AGPS)
+#include <net/nrf_cloud_agps.h>
+#endif
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+#include <net/nrf_cloud_pgps.h>
+#endif
+
+BUILD_ASSERT(
+	IS_ENABLED(CONFIG_NRF_CLOUD_MQTT) &&
+	IS_ENABLED(CONFIG_NRF_CLOUD_CONNECTION_POLL_THREAD));
+
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+static struct k_work notify_pgps_work;
+#endif
+
+static const char cloud_shell_cmd_usage_str[] =
+	"MQTT connection to nRFCloud\n"
+	"Usage: cloud <subcommand>\n"
+	"\n"
+	"Subcommands:\n"
+	"  connect:      Establish MQTT connection to nRFCloud.\n"
+	"  disconnect:   Disconnect from nRFCloud.\n";
+
+static void cloud_shell_print_usage(const struct shell *shell)
+{
+	mosh_print_no_format(cloud_shell_cmd_usage_str);
+}
+
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+static void notify_pgps(struct k_work *work)
+{
+	ARG_UNUSED(work);
+	int err;
+
+	err = nrf_cloud_pgps_notify_prediction();
+	if (err) {
+		mosh_error("Error requesting notification of prediction availability: %d", err);
+	}
+}
+#endif /* defined(CONFIG_NRF_CLOUD_PGPS) */
+
+static void nrf_cloud_event_handler(const struct nrf_cloud_evt *evt)
+{
+	int err = 0;
+
+	switch (evt->type) {
+	case NRF_CLOUD_EVT_TRANSPORT_CONNECTING:
+		mosh_print("NRF_CLOUD_EVT_TRANSPORT_CONNECTING");
+		break;
+	case NRF_CLOUD_EVT_TRANSPORT_CONNECTED:
+		mosh_print("NRF_CLOUD_EVT_TRANSPORT_CONNECTED");
+		break;
+	case NRF_CLOUD_EVT_READY:
+		mosh_print("NRF_CLOUD_EVT_READY");
+		break;
+	case NRF_CLOUD_EVT_TRANSPORT_DISCONNECTED:
+		mosh_print("NRF_CLOUD_EVT_TRANSPORT_DISCONNECTED");
+		break;
+	case NRF_CLOUD_EVT_ERROR:
+		mosh_print("NRF_CLOUD_EVT_ERROR");
+		break;
+	case NRF_CLOUD_EVT_SENSOR_DATA_ACK:
+		mosh_print("NRF_CLOUD_EVT_SENSOR_DATA_ACK");
+		break;
+	case NRF_CLOUD_EVT_FOTA_DONE:
+		mosh_print("NRF_CLOUD_EVT_FOTA_DONE");
+		break;
+	case NRF_CLOUD_EVT_RX_DATA:
+		mosh_print("NRF_CLOUD_EVT_RX_DATA");
+		if (((char *)evt->data.ptr)[0] == '{') {
+			/* Not A-GPS data */
+			break;
+		}
+#if defined(CONFIG_NRF_CLOUD_AGPS)
+		err = nrf_cloud_agps_process((char *)evt->data.ptr, evt->data.len);
+		if (!err) {
+			mosh_print("A-GPS data processed");
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+			/* call us back when prediction is ready */
+			k_work_submit(&notify_pgps_work);
+#endif
+			/* data was valid; no need to pass to other handlers */
+			break;
+		}
+#endif
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+		err = nrf_cloud_pgps_process(evt->data.ptr, evt->data.len);
+		if (err) {
+			mosh_error("Error processing P-GPS packet: %d", err);
+		}
+#else
+		if (err) {
+			mosh_print("Unable to process A-GPS data, error: %d", err);
+		}
+#endif
+		break;
+	case NRF_CLOUD_EVT_USER_ASSOCIATION_REQUEST:
+		mosh_print("NRF_CLOUD_EVT_USER_ASSOCIATION_REQUEST");
+		mosh_warn("Add the device to nRF Cloud and wait for it to reconnect");
```

Comment:
Does the nRF Cloud library trigger reconnect automatically or is this something that the application needs to do?

Reply:
Maybe not, but the warning trace kind of suggests that it happens automatically. :)

Reply:
You're right, I may have used copy-paste a bit too liberally here :) I'll change the warning message for now. If there is a case for having auto-reconnect, that can be implemented too.